### PR TITLE
Feature #143 GPGS for Profile

### PR DESCRIPTION
**Pull request by EugeneButusov**

### CODE REVIEW

File: SoomlaAndroidProfile/src/com/soomla/profile/SoomlaProfile.java [hunk 1]
```diff
@@ -1152,6 +1158,170 @@ public void multiShare(String text, String imageFilePath) {
...
+     * @param payload  a String to receive when the function returns.
+     * @param reward   The reward to grant
+     * @throws ProviderNotFoundException if the supplied provider is not
+     *                                   supported by the framework
+     */
+    public void getScores(final IProvider.Provider provider, final Leaderboard leaderboard, final boolean fromStart, final String payload, final Reward reward) throws ProviderNotFoundException {
+        final IGameServicesProvider gsProvider = mProviderManager.getGameServicesProvider(provider);
+
+        BusProvider.getInstance().post(new GetScoresStartedEvent(provider, leaderboard, fromStart, payload));
+        gsProvider.getScores(leaderboard.getID(), fromStart, new GameServicesCallbacks.SuccessWithListListener<Score>() {
```
Comment (vedi) [inline SoomlaAndroidProfile/src/com/soomla/profile/SoomlaProfile.java]: @EugeneButusov, sorry I missed this before...
`getID()` breaks naming conventions in Java, change it to `getId()`

Comment (vedi) [inline SoomlaAndroidProfile/src/com/soomla/profile/SoomlaProfile.java]: oh, if it goes from Soomla base classes, keep it as is :)

File: SoomlaAndroidProfile/src/com/soomla/profile/domain/gameservices/Leaderboard.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * Constructor
+     *
+     * @param ID the leaderboard ID for the given provider
+     * @param provider the provider which the user's data is associated to
+     * @param name the name of this leaderboard
+     * @param iconUrl url to icon of this leaderboard
+     */
+    public Leaderboard(String ID, IProvider.Provider provider, String name, String iconUrl) {
+        mID = ID;
```
Comment (vedi) [inline SoomlaAndroidProfile/src/com/soomla/profile/domain/gameservices/Leaderboard.java]: change ID -> id

File: SoomlaAndroidProfile/src/com/soomla/profile/domain/gameservices/Leaderboard.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        return mName;
+    }
+
+    public String getIconURL() {
+        return mIconUrl;
+    }
+
+    /** Private Members **/
+
+    private String mID;
```
Comment (vedi) [inline SoomlaAndroidProfile/src/com/soomla/profile/domain/gameservices/Leaderboard.java]: mID -> mId

File: SoomlaAndroidProfile/src/com/soomla/profile/events/gameservices/GetLeaderboardsFailedEvent.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+package com.soomla.profile.events.gameservices;
+
+import com.soomla.profile.domain.IProvider;
+
+public class GetLeaderboardsFailedEvent extends BaseGameServicesEvent {
+    public final String ErrorDescription;
+
+    public GetLeaderboardsFailedEvent(IProvider.Provider provider, String errorDescription, String payload) {
+        super(provider, payload);
+        this.ErrorDescription = errorDescription;
```
Comment (vedi) [inline SoomlaAndroidProfile/src/com/soomla/profile/events/gameservices/GetLeaderboardsFailedEvent.java:26]: @EugeneButusov, be careful `ErrorDescription` breaks naming conventions in Java. Soomla does that in their classes, but I'm not sure you should do the same. Follow case by case.

Comment (vedi) [inline SoomlaAndroidProfile/src/com/soomla/profile/events/gameservices/GetLeaderboardsFailedEvent.java:26]: Actually there are 10s such cases below.

File: social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java [hunk 5]
```diff
@@ -81,12 +89,14 @@
     public static final int ACTION_PUBLISH_STATUS_DIALOG = 4;
 
     private boolean autoLogin;
+    private boolean enableGameServices;
 
     private String lastContactCursor = null;
+    private HashMap<String, LeaderboardScoreBuffer> scoresCursors = null;
```
Comment (vedi) [inline social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java]: @EugeneButusov, you should not use an exact class here, change to `Map<String, LeaderboardScoreBuffer>`

File: social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java [hunk 6]
```diff
@@ -492,6 +516,136 @@ public void getFeed(Boolean fromStart, SocialCallbacks.FeedListener feedsListene
...
+        if (enableGameServices) {
+            LeaderboardScoreBuffer leaderboardScores = scoresCursors.get(leaderboardId);
+            if (fromStart || leaderboardScores == null) {
+                Games.Leaderboards.loadTopScores(googleApiClient, leaderboardId, LeaderboardVariant.TIME_SPAN_ALL_TIME, LeaderboardVariant.COLLECTION_PUBLIC, ITEMS_PER_PAGE)
+                        .setResultCallback(new ResultCallback<Leaderboards.LoadScoresResult>() {
+                            @Override
+                            public void onResult(Leaderboards.LoadScoresResult loadScoresResult) {
+                                if (loadScoresResult.getStatus().isSuccess()) {
+                                    scoresCursors.put(leaderboardId, loadScoresResult.getScores());
+                                    ArrayList<Score> result = new ArrayList<Score>();
```
Comment (vedi) [inline social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java]: Do not use an exact class in declaration, change to `List<Score>`.

Comment (vedi) [inline social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java]: It's general rule.

File: SoomlaAndroidProfile/src/com/soomla/profile/SoomlaProfile.java [hunk 7]
```diff
@@ -1152,6 +1158,170 @@ public void multiShare(String text, String imageFilePath) {
         SoomlaApp.getAppContext().startActivity(chooser);
     }
 
+    /**
+     Fetches the game's leaderboards list
+
+     * @param provider The provider to use
```
Comment (vedi) [inline SoomlaAndroidProfile/src/com/soomla/profile/SoomlaProfile.java:1164]: You missed "*" in the beginning of couple lines. And in other places below.

File: social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java (hunk 6 above)
Comment (EugeneButusov) [inline social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java]: @vedi, fixed.

File: social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java (hunk 5 above)
Comment (EugeneButusov) [inline social-providers/android-profile-google/src/com/soomla/profile/social/google/SoomlaGooglePlus.java]: @vedi, done.

File: SoomlaAndroidProfile/src/com/soomla/profile/events/gameservices/GetLeaderboardsFailedEvent.java (hunk 4 above)
Comment (EugeneButusov) [inline SoomlaAndroidProfile/src/com/soomla/profile/events/gameservices/GetLeaderboardsFailedEvent.java:26]: @vedi, I think it should follow one code style. Maybe it's better to make in separate ticket and refactor it everywhere.

File: SoomlaAndroidProfile/src/com/soomla/profile/SoomlaProfile.java (hunk 1 above)
Comment (EugeneButusov) [inline SoomlaAndroidProfile/src/com/soomla/profile/SoomlaProfile.java]: @vedi, Leaderboard hasn't superclass, I'll refactor as you said.

**Pull request merged**